lib/portlayer/attach/attach.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package attach
+
+import (
+	"fmt"
+	"net"
+
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25/types"
+	"github.com/vmware/vic/lib/portlayer/constants"
+	"github.com/vmware/vic/lib/portlayer/exec"
+	"github.com/vmware/vic/pkg/trace"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+// Handler defines the interface
+type Handler interface {
+	Join(handle interface{}) (interface{}, error)
+	Bind(handle interface{}) (interface{}, error)
+	Unbind(handle interface{}) (interface{}, error)
+}
+
+func toggle(handle *exec.Handle, connected bool) (*exec.Handle, error) {
+	// make sure a spec exists
+	handle.SetSpec(nil)
+	// get the virtual device list
+	devices := object.VirtualDeviceList(handle.Container.Config.Hardware.Device)
+
+	// select the virtual serial ports
+	serials := devices.SelectByBackingInfo((*types.VirtualSerialPortURIBackingInfo)(nil))
+	if len(serials) == 0 {
+		return nil, fmt.Errorf("Unable to bind a device with desired backing")
+	}
+	if len(serials) > 1 {
+		return nil, fmt.Errorf("Multiple matches found with desired backing")
+	}
+
+	for i := range serials {
+		serial := serials[i]
+		log.Debugf("Found a device with desired backing: %#v", serial)
+
+		if serial.GetVirtualDevice().Connectable.Connected == connected {
+			log.Debugf("Already in the desired state (connected: %t)", connected)
+			return handle, nil
+		}
+
+		log.Debugf("Setting the connected state to %t", connected)
+		// set the values
+		serial.GetVirtualDevice().Connectable.Connected = connected
+		if connected && handle.ExecConfig.Sessions[handle.ExecConfig.ID].Attach {
```

Comment:
This will almost always start the container connected. The only time that's not the case is with `run -d`.
This should be addressed in the follow up PR we're been discussing that adds run-blocking.